typing_extensions/src_py2/typing_extensions.py

```diff
@@ -89,3 +95,259 @@ def __subclasshook__(cls, C):
+        bases = tuple(b._gorg if isinstance(b, GenericMeta) else b for b in bases)
+
+        if any(isinstance(b, GenericMeta) and b is not Generic for b in bases):
+            bases = tuple(b for b in bases if b is not Generic)
+        namespace.update({'__origin__': origin, '__extra__': extra})
+        self = abc.ABCMeta.__new__(cls, name, bases, namespace)
+        abc.ABCMeta.__setattr__(self, '_gorg', self if not origin else origin._gorg)
+
+        self.__parameters__ = tvars
+        self.__args__ = tuple(Ellipsis if a is _TypingEllipsis else
+                              () if a is _TypingEmpty else
+                              a for a in args) if args else None
+        self.__next_in_mro__ = _next_in_mro(self)
+        if orig_bases is None:
+            self.__orig_bases__ = initial_bases
+        if (
+            '__subclasshook__' not in namespace and extra or
+            getattr(self.__subclasshook__, '__name__', '') == '__extrahook__'
+        ):
+            self.__subclasshook__ = _make_subclasshook(self)
+        self.__tree_hash__ = (hash(self._subs_tree()) if origin else
+                              abc.ABCMeta.__hash__(self))
+        return self
+
+    def __init__(cls, *args, **kwargs):
+        super(_ProtocolMeta, cls).__init__(*args, **kwargs)
+        if not cls.__dict__.get('_is_protocol', None):
+            cls._is_protocol = any(b is Protocol or
+                                   isinstance(b, _ProtocolMeta) and
+                                   b.__origin__ is Protocol
+                                   for b in cls.__bases__)
+        if cls._is_protocol:
+            for base in cls.__mro__[1:]:
+                if not (base in (object, Generic, Callable) or
+                        isinstance(base, TypingMeta) and base._is_protocol or
+                        isinstance(base, GenericMeta) and base.__origin__ is Generic or
+                        _collection_protocol(base)):
+                    raise TypeError('Protocols can only inherit from other protocols,'
+                                    ' got %r' % base)
+
+            def _no_init(self, *args, **kwargs):
+                if type(self)._is_protocol:
+                    raise TypeError('Protocols cannot be instantiated')
+            cls.__init__ = _no_init
+
+        def _proto_hook(cls, other):
+            if not cls.__dict__.get('_is_protocol', None):
+                return NotImplemented
+            for attr in cls._get_protocol_attrs():
+                for base in other.__mro__:
+                    if attr in base.__dict__:
+                        if base.__dict__[attr] is None:
+                            return NotImplemented
+                        break
+                else:
+                    return NotImplemented
+            return True
+        if '__subclasshook__' not in cls.__dict__:
+            cls.__subclasshook__ = classmethod(_proto_hook)
+
+    def __instancecheck__(self, instance):
```

Comment:
This maybe a glitch in the logic of treating ``None``. The ``None`` is special, since ``__iter__ = None``, explicitly indicates that a class is _not_ iterable at runtime. I will double check.

Reply:
This doesn't seem quite right. Here is another Python 2 example (Python 3 is similar):

```py
from typing import Optional
from typing_extensions import Protocol, runtime

@runtime
class P(Protocol):
    x = None  # type: Optional[str]

class B(object):
    def __init__(self):  # type: () -> None
        self.x = None  # type: Optional[str]

b = B()
print isinstance(b, P) # False
b.x = ''
print isinstance(b, P) # True
b.x = None
print isinstance(b, P) # False
```

I'd expect the output to be `True`/`True`/`True`. What about only using the special `None` logic for protocol members that are callables in the protocol? They could still be used to represent a missing method. `None` is a valid value for a non-method attribute, so special casing those cases seems wrong to me.

(Overriding with `None` breaks Liskov substitutability but supporting it is reasonable since it's a Python idiom, for better or worse.)

Reply:
> What about only using the special None logic for protocol members that are callables in the protocol?

What do you think about doing the same that mypy currently does -- apply the ``None`` in subclass rule only to dunder attributes? (like ``__iter__``). Maybe even better idea would be to do _exactly_ the same as ABCs in ``collections.abc`` do, i.e. only check whether e.g. ``__iter__`` is set to ``None`` on class, not on instances. What do you think?

Reply:
Just checking for `None` attribute in the class sounds like a good idea, thought it wouldn't be enough by itself. For example, `B` could be defined like:

```py
class B(object):
    x = None  # type: Optional[str]
```

Restriction to dunder attributes seems too ad hoc -- the `None` check is also used for non-dunder methods such as `send` and `throw` in `collections.abc` (https://github.com/python/cpython/blob/master/Lib/_collections_abc.py#L151). Mypy probably shouldn't use it either.

What about doing a `callable()` check on the attribute value defined in the protocol and requiring that the type object of the instance has a `None` value for the corresponding attribute? I think that this would be unlikely to cause problems and would cover all important use cases. It's still possible to construct an example which doesn't work right, but that's acceptable since `@runtime` was never supposed to be 100% correct anyway.

Reply:
> What about doing a callable() check on the attribute value defined in the protocol and requiring that the type object of the instance has a None value for the corresponding attribute?

OK, I will go with this (applying the ``None`` rule only to callable attributes, and check them on the class object). I will also open an issue in mypy, to do the same (instead of the ad-hoc dunder restriction).